### ISSUE TITLE
ArrayOutOfBound exception in random generation

### ISSUE THREAD

**Issue by lesquoyb** [gama-platform/gama.old#3572]
**Describe the bug**
In a gama model, I was generating a huge matrix with random data and it raised an out of bound exception in the code of the Mersenne random generator. Unfortunately I didn't set the seed explicitly so I'm not able to reproduce it, but here is the stack trace I got in eclipse console:
```
java.lang.ArrayIndexOutOfBoundsException: Index 624 out of bounds for length 624
	at msi.gama.util.random.MersenneTwisterRNG.next(MersenneTwisterRNG.java:216)
	at java.base/java.util.Random.nextDouble(Random.java:463)
	at msi.gama.common.util.RandomUtils.next(RandomUtils.java:500)
	at msi.gama.common.util.RandomUtils.between(RandomUtils.java:444)
	at msi.gaml.operators.Colors.random_color(Colors.java:588)
	at gaml.additions.core.GamlAdditions.lambda$397(GamlAdditions.java:455)
	at msi.gaml.expressions.operators.UnaryOperator._value(UnaryOperator.java:105)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:81)
	at msi.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:91)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:81)
	at msi.gaml.types.GamaMatrixType.lambda$3(GamaMatrixType.java:205)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfInt.accept(ForEachOps.java:204)
	at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
	at java.base/java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:711)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

And here is the gaml code that produced it:
```
global skills:[thread] {
	matrix<int> big_matrix;
	
	init {
		big_matrix <- point(1000,1000) matrix_with int(rnd_color(255));
        do run_thread;
	}
    action thread_action {
	    big_matrix <- point(width,height) matrix_with int(rnd_color(255));
    }
}
```
I don't think it is linked to the `thread` skill at all, but leaving it for context.
I attached the full console output for context too: [full console log.txt](https://github.com/gama-platform/gama/files/10349099/full.console.log.txt)
Not sure it deserves an issue as I can't reproduce it yet, but I'll try to investigate more later

**Comment by lesquoyb** [comment]
Some updates: 
Adding a try/catch statement around the faulty code in MersenneTwisterRNG gave me surprising results: I am now able to catch the bug everytime I run the model while before it only happened now and then, and was not reproducible even when I caught the seed used.
What's more surprising is that it happens multiple times per simulation and everytime, the issue is the same: ArrayOutOfBound, but the actual index used is not out of bound at all.
All of that and some discussion with @ptaillandier lead me to think that it is related to the use of threads. 
We can actually remove the `thread` skill from the model and still observe the problem:
```
model threadbug

global {
	
	matrix<int> big_matrix;
	int width <- 1000;
	int height <- 1000;
	init {
		big_matrix <- point(width,height) matrix_with int(rnd_color(255));
	}
}


experiment test {

}
```
You will need to change the line 220:
```
y = mt[mtIndex++];
```
In `MersenneTwisterRNG.java`, with:
```
try {
	y = mt[mtIndex++];			
}
catch (Exception e) {
	DEBUG.OUT(e);
	throw e;
}
```
To have the errors actually showed in the gama interface (one more clue that it was a threading related error that went silent).

So, taking into account that it looks like a threading error, I think the problem comes from the `matrix_with` operator, that seems to run the random generator in a separate thread for every cell.

**Comment by AlexisDrogoul** [comment]
Yep. `matrix_with` tries to feed the matrix in parallel when concurrency is available. And it is highly probable that running the RNG in such a way can lead to unforeseen problems (the RNG we use is not thread-safe). Probably the fix is to revert to a sequential initialisation, what do you think ?

**Comment by lesquoyb** [comment]
Maybe we can add a boolean facet to run it in parallel or not. I definitely can see the advantages of it. We can even add an info to inform the user that it shouldn't be initialized with random processes in case the parallel option is set to true ?

**Comment by AlexisDrogoul** [comment]
Given that #3576 is also affected by this, I think we could imagine to propose a parallel RNG (specifying that others are not supposed to work in parallel), thanks to the new random generators brought by Java 17. And maybe set it by default when parallel is true ?

**Comment by lesquoyb** [comment]
Definitely a good idea yes

**Comment by AlexisDrogoul** [comment]
Have you tried setting the rng to `java`, by the way ? It might provide a temporary workaround to these issues.

**Comment by AlexisDrogoul** [comment]
I have committed (in 7cd7843df98db6c0702ca11ab7ba65dd50a2ee6d) a number of changes to fix this issue as well as provide more reliable algorithms for random number generation, especially in concurrent environments. 

1) Removal of the "cellular" algorithm, barely used, too simple, and not thread safe.
2) Reorganisation of the classes for better readability and compatibility with Java RandomGenerator.
3) Addition of two thread safe algorithms : `parallel`, a mersenne twister with a reentrant lock to prevent concurrent stage change -- it remains deterministic at the expense of implicitly synchronising all the threads accessing it (at least the ones accessing the critical section) ; `threaded`, a RNG based on `ThreadLocalRandom`, very fast even in highly concurrent, but that cannot be seedable (and therefore is not deterministic). The first one is probably a good choice for parallel simulations that try to keep some determinism; the second one for interactive parallel simulations.
4) Better documentation for these algorithms in GAML. 

I did not implement any "automatic" switch to a thread-safe algorithm when in parallel, as the choice of one will really depend on the scenario. Only thing we could is to make the error message a bit more explicit when the RNG fails ...
